src/lib/timing.c

```diff
@@ -0,0 +1,230 @@
+/*
+   +----------------------------------------------------------------------+
+   | Xdebug                                                               |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2002-2020 Derick Rethans                               |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 1.01 of the Xdebug license,   |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available at through the world-wide-web at                           |
+   | https://xdebug.org/license.php                                       |
+   | If you did not receive a copy of the Xdebug license and are unable   |
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+   | Authors: Derick Rethans <derick@xdebug.org>                          |
+   |          Michael Voříšek <mvorisek@mvorisek.cz>                      |
+   +----------------------------------------------------------------------+
+ */
+
+#include "php_xdebug.h"
+#if PHP_WIN32
+# include "win32/time.h"
+# include <versionhelpers.h>
+#else
+# include <sys/time.h>
+#endif
+#if __APPLE__
+# include <mach/mach_time.h>
+#endif
+
+#include "timing.h"
+
+#define NANOTIME_MIN_STEP 10
+
+#if PHP_WIN32
+# define WIN_NANOS_IN_TICK          100
+# define WIN_TICKS_SINCE_1601_JAN_1 116444736000000000ULL
+#endif
+
+ZEND_EXTERN_MODULE_GLOBALS(xdebug)
+
+// Windows
+#if PHP_WIN32
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
+{
+	FILETIME filetime;
+
+	if (nanotime_context->win_precise_time_func != NULL) {
+		nanotime_context->win_precise_time_func(&filetime);
+		return ((((uint64_t)filetime.dwHighDateTime << 32) + (uint64_t)filetime.dwLowDateTime) - WIN_TICKS_SINCE_1601_JAN_1)
+			* WIN_NANOS_IN_TICK;
+	}
+
+	return 0;
+}
+
+// Linux/Unix with clock_gettime
+#elif HAVE_CLOCK_GETTIME
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
+{
+   struct timespec ts;
+
+#ifdef CLOCK_MONOTONIC
```

Comment:
it is compile time check - when does it help, are there a platforms without a `CLOCK_MONOTONIC`?

Reply:
Yeah, it's possible. But after re-reading https://stackoverflow.com/a/50245503/508057 again I've adjusted the code a little bit so that it always falls back to `CLOCK_REALTIME`.